apps/extension/src/approve-origin.ts

```diff
@@ -0,0 +1,44 @@
+import { JsonValue } from '@bufbuild/protobuf';
+import { ConnectError } from '@connectrpc/connect';
+import { errorFromJson } from '@connectrpc/connect/protocol-connect';
+import { localExtStorage } from '@penumbra-zone/storage';
+import { OriginApproval, PopupType } from './message/popup';
+import { popup } from './popup';
+
+export const originAlreadyApproved = async (url: string): Promise<boolean> => {
+  // parses the origin and returns a consistent format
+  const urlOrigin = new URL(url).origin;
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  const connectedSites = (await localExtStorage.get('connectedSites')) ?? {};
+  return Boolean(connectedSites[urlOrigin]);
+};
+
+export const approveOrigin = async ({
+  origin: senderOrigin,
+  tab,
+  frameId,
+}: chrome.runtime.MessageSender): Promise<boolean> => {
+  if (!senderOrigin?.startsWith('https://') || !tab?.id || frameId)
+    throw new Error('Unsupported sender');
+
+  // parses the origin and returns a consistent format
+  const urlOrigin = new URL(senderOrigin).origin;
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  const connectedSites = (await localExtStorage.get('connectedSites')) ?? {};
```

Comment:
we do set a default, but that does not guarantee this will never fail. in the case where the storage object is somehow not present, failing to handle this creates an unrecoverable error. i'll change the storage type to account for this and enforce checks